### PR TITLE
Document RAND_DRBG fork-safety locking model

### PR DESCRIPTION
**Pull request by kaduk**
Add some more exposition on why unlocked access to the global rand_fork_count
is safe, and provide a comment for the struct rand_drbg_st fork_count field.

### CODE REVIEW

File: crypto/rand/rand_lcl.h [hunk 1]
```diff
@@ -88,7 +88,7 @@ struct rand_drbg_st {
     CRYPTO_RWLOCK *lock;
     RAND_DRBG *parent;
     int nid; /* the underlying algorithm */
-    int fork_count;
+    int fork_count; /* snapshot of global state when we last reseeded */
```
Comment (richsalz) [inline crypto/rand/rand_lcl.h]: I find this comment hard to parse.  I normally wouldn't care, but since the point of this PR is to make thngs more clear by adding comments the proposal isn't good.

File: crypto/rand/rand_lcl.h [hunk 2]
```diff
@@ -128,7 +128,12 @@ extern RAND_BYTES_BUFFER rand_bytes;
 extern RAND_DRBG rand_drbg;
 extern RAND_DRBG priv_drbg;
 
-/* How often we've forked (only incremented in child). */
+/*
+ * A "generation count" of forks.  Increment-only, and only ever written to
+ * in the child process of the fork, which is guaranteed to be single-threaded.
+ * As such, no locking is needed for normal (read) accesses, since the rest of
+ * fork processing is assumed to introduce the necessary memory barriers.
```
Comment (richsalz) [inline crypto/rand/rand_lcl.h]: Incorrect -- it's not the fork/syscall processing it's the pthreads implementation that handles it.

File: crypto/rand/rand_lcl.h (hunk 1 above)
Comment (kaduk) [inline crypto/rand/rand_lcl.h]: Okay.

File: crypto/rand/rand_lcl.h (hunk 2 above)
Comment (kaduk) [inline crypto/rand/rand_lcl.h]: My thinking was that the pthread library's fork handling is part of the broader "fork processing" umbrella, and the pthread implementation has to get pretty intertwined with libc anyway, but I can reword this if you want.

Comment (dot-asm) [inline crypto/rand/rand_lcl.h]: Is it insignificant that that both children will get same count?

Comment (richsalz) [inline crypto/rand/rand_lcl.h]: Yes it is not significant.  As long as the "number I've times I've been a child" doesn't match "number of times the process has created a child" then we reseed.

**Comment by richsalz** [comment]
I am not a fan because, as this PR so far has shown, it is too hard to document correctly.

**Comment by kaduk** [comment]
> I am not a fan because, as this PR so far has shown, it is too hard to document correctly.

Even if it is hard to document correctly (which I'm not fully convinced of yet), it's also hard to *implement* correctly, which seems an even more important concern.  Documenting how the implementation works and why it's safe lets future readers understand the code without having to redo the thread-safety analysis, and also helps protect against potential future modifications that invalidate the locking model.

**Comment by richsalz** [comment]
Shrug.  We disagree.

**Comment by kaduk** [comment]
I guess we are waiting for another OMC opinion about whether I should go forward with this or abandon it.

**Comment by kaduk** [comment]
rebased+squashed.  I am not sure whether Rich wanted additional text change to the "rest of fork processing" text.

**Comment by richsalz** [comment]
I'll stay out of this one now, we have other committers with more expertise.

File: crypto/rand/rand_lcl.h [hunk 3]
```diff
@@ -117,6 +117,10 @@ struct rand_drbg_st {
     RAND_DRBG *parent;
     int secure; /* 1: allocated on the secure heap, 0: otherwise */
     int nid; /* the underlying algorithm */
+    /*
+     * The value of the rand_fork_count global as of when we last reseeded.
+     * Used to provide fork-safety and reseed this DRBG in the child.
+     */
```
Comment (mspncp) [inline crypto/rand/rand_lcl.h]: Just a suggestion for a minor enhancement. Not a must-do.
``` 
/*
 * Stores the value of the rand_fork_count global as of when we last reseeded.
 * The DRG reseeds automatically whenever drbg->fork_count != rand_fork_count.
 * Used to provide fork-safety and reseed this DRBG in the child process.
 */
```

File: crypto/rand/rand_lcl.h [hunk 4]
```diff
@@ -203,7 +207,16 @@ struct rand_drbg_st {
 /* The global RAND method, and the global buffer and DRBG instance. */
 extern RAND_METHOD rand_meth;
 
-/* How often we've forked (only incremented in child). */
+/*
+ * A "generation count" of forks.  Increment-only, and only ever written to
```
Comment (mspncp) [inline crypto/rand/rand_lcl.h]: Your focus is so much on the fork safety that you 'forget' to state the essential:

```
* A "generation count" of forks. Incremented in the child process after a fork.
*
* Since rand_fork_ount is increment-only, and only ever written to ...
```

File: crypto/rand/rand_lcl.h (hunk 2 above)
Comment (mspncp) [reply crypto/rand/rand_lcl.h]: How about simply writing "pthread fork processing" instead of "fork processing"?

File: crypto/rand/rand_lcl.h (hunk 3 above)
Comment (levitte) [reply crypto/rand/rand_lcl.h]: I would rather see that enhancement.

File: crypto/rand/rand_lcl.h (hunk 2 above)
Comment (levitte) [reply crypto/rand/rand_lcl.h]: Sounds good to me

File: crypto/rand/rand_lcl.h (hunk 4 above)
Comment (levitte) [reply crypto/rand/rand_lcl.h]: I would like to see that enhancement too.

**Comment by mattcaswell** [comment]
@kaduk looks like we are waiting on your response to the raised issues.

**Comment by kaduk** [comment]
I'm happy to take the enhancements, and have just been busy and not gotten around to updating this yet.
(I re-wrapped the first one, which was at 82 columns.)

**Comment by kaduk** [comment]
Also rebased past the nid-->type conversion.

File: crypto/rand/rand_lcl.h [hunk 5]
```diff
@@ -203,7 +209,17 @@ struct rand_drbg_st {
 /* The global RAND method, and the global buffer and DRBG instance. */
 extern RAND_METHOD rand_meth;
 
-/* How often we've forked (only incremented in child). */
+/*
+ * A "generation count" of forks.  Incremented in the child process after a
+ * fork.  Since rand_fork_count is increment-only, and only ever written to in
+ * the child process of the fork, which is guaranteed to be single-threaded, no
+ * locking is needed for normal (read) accesses; the rest of fork processing is
```
Comment (mspncp) [inline crypto/rand/rand_lcl.h]: What do you think about [my suggestion](https://github.com/openssl/openssl/pull/4110#discussion_r172009849) to write "pthread fork processing" instead of "fork processing" to please @richsalz (favoured also by @levitte)?

**Comment by mattcaswell** [comment]
My approval stands with or without @mspncp's suggested tweak.

File: crypto/rand/rand_lcl.h (hunk 5 above)
Comment (kaduk) [reply crypto/rand/rand_lcl.h]: That is a good change; sorry for missing it in my hasty review of the outstanding comments.

**Comment by mspncp** [comment]
Mine too ;-)

**Comment by levitte** [comment]
So, errr, is anyone gonna merge this?  ;-)

**Comment by mspncp** [comment]
I thought for committers it's DIY ;-). But if Ben is too busy, I can merge it this evening.

**Comment by kaduk** [comment]
I was too busy last week, but I can do it now.

**Comment by kaduk** [comment]
Merged; thanks for the reminder!
